src/plan/generational/barrier.rs

```diff
@@ -0,0 +1,158 @@
+//! Generational read/write barrier implementations.
+
+use atomic::Ordering;
+
+use crate::plan::barriers::Barrier;
+use crate::policy::space::Space;
+use crate::scheduler::WorkBucketStage;
+use crate::util::metadata::compare_exchange_metadata;
+use crate::util::metadata::load_metadata;
+use crate::util::*;
+use crate::vm::{ObjectModel, VMBinding};
+use crate::MMTK;
+
+use super::gc_work::GenNurseryProcessEdges;
+use super::gc_work::ProcessArrayCopyModBuf;
+use super::gc_work::ProcessModBuf;
+use super::global::Gen;
+
+/// Object barrier for generational collection
+pub struct GenObjectBarrier<VM: VMBinding> {
```

Comment:
Is there any specific reason to move `ObjectBarrier` to `GenObjectBarrier`?

Reply:
This is not specific to a plan, but a category of plans. Here `GenObjectBarrier` is used for generational plans and is shared by GenCopy and GenImmix. 

I moved `ObjectBarrier` to `GenObjectBarrier` because our implementation of the object barrier is only used for generational collection. Object barrier itself can be used for other purposes, including reference counting and concurrent marking, and they're not supported by this implementation. So it is more correct to rename it to `GenObjectBarrier` and only use it for generational plans.

Reply:
One of our principles is to encourage reusable component. Unless we agree that write barriers are not reusable, this implementation is against our design.

Reply:
It is reusable. It's already shared by `GenCopy` and `GenImmix`. This `GenObjectBarrier` is not bounded to a specific plan.

> What if someone would use an object barrier for other plans?

This depends on the actual usage. If the new plan wants to use it for generational collection, it should directly reuse `GenObjectBarrier`. If it is used for other purposes like concurrent marking, (1) notice that the old `ObjectBarrier` cannot handle this case as well, and cannot be reused directly; (2) the plan should create another `SATBObjectBarrier` for concurrent marking.

Reply:
With this implementation, `SATBObjectBarrier` and `GenObjectBarrier` in your example cannot share any barrier code, even if they both use/are object barriers. That is the problem.

From `ObjectBarrier` to `GenObjectBarrier`, it is not a simple renaming. They are different in a number of ways:
* `GenObjectBarrier` has an array copy barrier other than the object barrier.
* `ObjectBarrier` does not assume a specific plan (or 'a category of plans'), and `GenObjectBarrier` assumes `Gen`.
* `ObjectBarrier` does not assume which metadata it is using, `GenObjectBarrier` assumes it uses the log bit.

Let's discuss this in tomorrow's meeting.

Reply:
Based on today's discussion, a barrier has both 'mechanism' and 'semantics'. The current `GenObjectBarrier` describes the semantics of the barrier, and we would extract the mechanism part to a separate type, and make it reusable.

Reply:
The object barrier type is now `ObjectBarrier<Semantics>`, which accepts a type argument specifying its slow-path semantics, and UNLOG_BIT metadata spec. The fast path code is defined in `ObjectBarrier` and can be shared by all the plans that may use this barrier. The slow-path code is defined in `Semantics`, and can only be shared by plans with same slow-path semantics.